src/main/java/com/gregtechceu/gtceu/api/machine/multiblock/MultiblockControllerMachine.java

```diff
@@ -204,4 +218,78 @@ public void onRotated(Direction oldFacing, Direction newFacing) {
             mwsd.addAsyncLogic(this);
         }
     }
+
+    public boolean allowExtendedFacing() {
+        return getDefinition().isAllowExtendedFacing();
+    }
+
+    public boolean allowFlip() {
+        return getDefinition().isAllowFlip();
+    }
+
+    @Override
+    public boolean isFacingValid(Direction facing) {
+        return allowExtendedFacing() || super.isFacingValid(facing);
+    }
+
+    public Direction getUpwardsFacing() {
+        return this.allowExtendedFacing() ? this.getBlockState().getValue(IMachineBlock.UPWARDS_FACING_PROPERTY) :
+                Direction.NORTH;
+    }
+
+    public void setUpwardsFacing(@NotNull Direction upwardsFacing) {
+        if (!getDefinition().isAllowExtendedFacing()) return;
+        if (upwardsFacing == null || upwardsFacing == Direction.UP || upwardsFacing == Direction.DOWN) {
+            GTCEu.LOGGER.error("Tried to set upwards facing to invalid facing {}! Skipping", upwardsFacing);
+            return;
+        }
+        BlockState blockState = getBlockState();
+        if (blockState.getBlock() instanceof MetaMachineBlock metaMachineBlock &&
+                blockState.getValue(IMachineBlock.UPWARDS_FACING_PROPERTY) != upwardsFacing) {
+            getLevel().setBlockAndUpdate(getPos(),
+                    blockState.setValue(IMachineBlock.UPWARDS_FACING_PROPERTY, upwardsFacing));
+            if (getLevel() != null && !getLevel().isClientSide) {
+                notifyBlockUpdate();
+                markDirty();
+                checkPattern();
+            }
+        }
```

Comment:
I'd suggest early returns here to make it a bit more readable with the long-ish conditions.

Something like this maybe... (please double check before applying lol)

```suggestion
        if (!(blockState.getBlock() instanceof MetaMachineBlock metaMachineBlock)) return;
        if (blockState.getValue(IMachineBlock.UPWARDS_FACING_PROPERTY) == upwardsFacing) return;
            
        getLevel().setBlockAndUpdate(
                getPos(),
                blockState.setValue(IMachineBlock.UPWARDS_FACING_PROPERTY, upwardsFacing)
        );
        
        if (getLevel() != null && !getLevel().isClientSide) {
            notifyBlockUpdate();
            markDirty();
            checkPattern();
        }
```